src/Castle.Facilities.AspNetCore/CastleWindsorRegistrationExtensions.cs

```diff
@@ -77,39 +208,68 @@ public static void UseCastleWindsor(this IApplicationBuilder app, IWindsorContai
 		/// <typeparam name="T">Type of service that implements Microsoft.AspNetCore.Http.IMiddleware</typeparam>
 		/// <param name="app">Application builder</param>
 		/// <param name="container">Windsor container</param>
-		public static void UseCastleWindsorMiddleware<T>(this IApplicationBuilder app, IWindsorContainer container) where T : class, IMiddleware
+		public static void UseMiddlewareFromWindsor<T>(this IApplicationBuilder app, IWindsorContainer container) where T : class, IMiddleware
 		{
 			container.Register(Component.For<T>());
 			app.Use(async (context, next) =>
 			{
 				var resolve = container.Resolve<T>();
-				await resolve.InvokeAsync(context, async (ctx) => await next());
+				try
+				{
+					await resolve.InvokeAsync(context, async (ctx) => await next());
+				}
+				finally
+				{
+					container.Release(resolve);
+				}
 			});
 		}
 
-		/// <summary>
-		/// Optional helpful exception to validate against types being installed into Windsor from the 'Microsoft.AspNetCore'
-		/// namespace. This guards against 'Torn Lifestyles' and 'Captive Dependencies' and will probably need to be extended.
-		/// </summary>
-		/// <param name="container">Windsor container</param>
-		public static void AssertNoAspNetCoreRegistrations(this IWindsorContainer container)
+		private static void AddApplicationComponentsToWindsor(IWindsorContainer container, CastleWindsorRegistrationOptions options)
 		{
-			var handlers = container.Kernel.GetHandlers();
-			var hasAspNetCoreFrameworkRegistrations = handlers.Any(x => x.ComponentModel.Implementation.Namespace.StartsWith("Microsoft.AspNetCore"));
-			if (hasAspNetCoreFrameworkRegistrations)
+			if (!options.ControllerRegistrations.Any())
+			{
+				container.Register(Classes.FromAssemblyInThisApplication(options.EntryAssembly).BasedOn<Controller>().LifestyleScoped());
+			}
+
+			foreach (var controllerRegistration in options.ControllerRegistrations)
 			{
-				throw new Exception(
-					"Looks like you have implementations registered from 'Microsoft.AspNetCore'. " +
-					"Please do not do this as it could lead to torn lifestyles and captive dependencies. " +
-					"Please remove the registrations from Castle.Windsor.");
+				container.Register(Classes.FromAssemblyInThisApplication(controllerRegistration.Item1).BasedOn<Controller>().Configure(x => x.LifeStyle.Is(controllerRegistration.Item2)));
+			}
+
+			if (!options.TagHelperRegistrations.Any())
+			{
+				container.Register(Classes.FromAssemblyInThisApplication(options.EntryAssembly).BasedOn<ViewComponent>().LifestyleScoped());
+			}
+
+			foreach (var controllerRegistration in options.TagHelperRegistrations)
+			{
+				container.Register(Classes.FromAssemblyInThisApplication(controllerRegistration.Item1).BasedOn<Controller>().Configure(x => x.LifeStyle.Is(controllerRegistration.Item2)));
```

Comment:
Should this and view components be based on a different class other than Controller, I think you had this right before you recent changes.

Reply:
And here come some tests. The history of this code is something that vexed me.

Reply:
Really vexed me.